### PR TITLE
case insensitive charset matching

### LINKED ISSUE #349: charset value handing does not follow http spec
in https://tools.ietf.org/html/rfc2616#section-3.4, 
> HTTP character sets are identified by case-insensitive tokens.

but in handler.rs, 
> 		match content_type.and_then(|val| val.to_str().ok()) {
>			Some("application/json") => true,
>			Some("application/json; charset=utf-8") => true,
only "utf-8" is considered as valid, and only one space is allowed. This causes lots of request failed.

### PR DESCRIPTION
**Pull request by seunlanlege**
closes #349

### CODE REVIEW

File: http/src/handler.rs [hunk 1]
```diff
@@ -562,10 +616,58 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
...
+	use jsonrpc_core::*;
+	use std::sync::atomic::{AtomicUsize};
+	
+	#[derive(Clone, Debug)]
+	struct Meta(usize);
+	
+	impl Metadata for Meta {}
+
+	#[derive(Default)]
+	struct MyMiddleware(AtomicUsize);
```
Comment (tomusdrw) [inline http/src/handler.rs]: Use `middleware::Noop`

File: http/src/handler.rs [hunk 2]
```diff
@@ -562,10 +616,58 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
...
 		}
 	}
 }
+
+#[cfg(test)]
+mod test {
+	use super::{hyper, RpcHandler};
+	use jsonrpc_core::futures::future::Either;
+	use jsonrpc_core::futures::Future;
+	use jsonrpc_core::*;
```
Comment (tomusdrw) [inline http/src/handler.rs]: Please avoid glob imports.

File: http/src/handler.rs [hunk 3]
```diff
@@ -562,10 +616,58 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
...
+			X: Future<Item = Option<Response>, Error = ()> + Send + 'static,
+		{
+			Either::A(Box::new(next(request, meta)))
+		}
+	}
+
+
+	#[test]
+	fn test_case_insensitive_content_type() {
+		let request = hyper::Request::builder()
```
Comment (tomusdrw) [inline http/src/handler.rs]: Can we please add another assertion to make sure that the builder doesn't normalize the value?

File: http/src/handler.rs [hunk 4]
```diff
@@ -562,10 +562,45 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
 	/// Returns true if the `content_type` header indicates a valid JSON
 	/// message.
 	fn is_json(content_type: Option<&header::HeaderValue>) -> bool {
-		match content_type.and_then(|val| val.to_str().ok()) {
-			Some("application/json") => true,
-			Some("application/json; charset=utf-8") => true,
+		match content_type.and_then(|val| val.to_str().map(str::to_lowercase).ok()) {
+			Some(ref content)
```
Comment (tomusdrw) [inline http/src/handler.rs]: Just noticed that we could avoid allocation of `to_lowercase` by just using `str::eq_ignore_ascii_case`.

File: http/src/handler.rs [hunk 5]
```diff
@@ -562,10 +562,45 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
...
+				if content == "application/json"
+					|| content == "application/json; charset=utf-8" =>
```
Comment (tomusdrw) [inline http/src/handler.rs]: Is the space relevant here? Should we match against no-space variant as well?

File: http/src/handler.rs [hunk 6]
```diff
@@ -562,10 +562,45 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
...
 }
+
+#[cfg(test)]
+mod test {
+	use super::{hyper, RpcHandler};
+	use jsonrpc_core::middleware::Noop;
+	use jsonrpc_core::Metadata;
+	
+	#[derive(Clone, Debug)]
+	struct Meta(usize);
```
Comment (tomusdrw) [inline http/src/handler.rs]: You can also use `()` instead of `Meta` instead of this dummy struct

File: http/src/handler.rs (hunk 5 above)
Comment (seunlanlege) [reply http/src/handler.rs]: the spec doesn't mention spaces, but to be on the safe side, let's match without spaces

File: http/src/handler.rs [hunk 7]
```diff
@@ -562,10 +562,39 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
 	/// Returns true if the `content_type` header indicates a valid JSON
 	/// message.
 	fn is_json(content_type: Option<&header::HeaderValue>) -> bool {
-		match content_type.and_then(|val| val.to_str().ok()) {
-			Some("application/json") => true,
-			Some("application/json; charset=utf-8") => true,
+		match content_type.and_then(|val| val.to_str().map(str::to_lowercase).ok()) {
+			Some(ref content)
+				if content.eq_ignore_ascii_case("application/json")
+					|| content.replace(" " ,"").eq_ignore_ascii_case("application/json;charset=utf-8") =>
```
Comment (tomusdrw) [inline http/src/handler.rs]: `replace` allocates a new string, let's duplicate the check maybe?

**Comment by leoaion** [comment]
Thanks for the quick fix. will you merge this fix to branch parity-2.2?

**Comment by leoaion** [comment]
will this fix be merged into parity-2.2 branch? or under any other released branch/tag?

**Comment by tomusdrw** [comment]
@leoaion yes, see backports here #350 later on it will get released as well.

**Pull request merged**